Refresh UX to enable HPOS tables

vedanshujain opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR makes the following changes to the feature-enabling flow for HPOS:

### Setting for specifying transactions is removed

Since many merchants would not be able to take an informed decision about which transaction setting to use (and whether to use transactions in the first place), this PR

1. Removes the setting to select the transaction and enabled it for everyone by default in `READ UNCOMMITTED` mode.
2. Reduces the scope of the transaction so that only insert and update statements of the batch are enclosed in a transaction. Currently, select statements (on the posts table) were also enclosed and therefore locks the post table for the duration of the transaction, which is something that is now avoided.

### Dedicated feature setting page is now folded into the feature controller page itself.

Currently, we have a dedicated feature setting page, which is now removed and folded into the features setting page itself so that merchant won't have to navigate to yet another screen to enable disable the feature.

### HPOS feature is elevated to stable and reduced the severity of warning messages

Currently, we have marked HPOS as experimental, however with this PR we have marked it as stable, and also reduced the severity of warning messages that we are showing.

So with this PR, the feature page will look like this:

When HPOS tables are authoritative

![Screenshot 2023-07-07 at 6 59 12 PM](https://github.com/woocommerce/woocommerce/assets/7571618/a9d0872b-3897-4967-8f54-d89e4a9152ba)

When post tables are authoritative (and incompatible plugins + un-synced orders are present)

![Screenshot 2023-07-07 at 6 57 32 PM](https://github.com/woocommerce/woocommerce/assets/7571618/41cc8941-9789-4102-827d-9792901c7564)




Closes #34975 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

**Pre-testing**
1. Start with a new show as this flow will be best tested when we start with a new shop on this branch. Alternatively, to start with the existing shop, switch to post tables, disable the sync, and drop the tables created for HPOS.
2. Create some products and some orders.
3. Add a dummy incompatible plugin. To do this create a new file called incompatible.php with the following contents and upload it to the site and activate the plugin
```php
<?php
/**
 * Plugin Name: Incompatible
 * WC requires at least: 5.0.0
 * WC tested up to: 7.4.0
 */

/**
 * Declare HPOS (in)compatibility.
 */
add_action( 'before_woocommerce_init', function() {
	if ( class_exists( \Automattic\WooCommerce\Utilities\FeaturesUtil::class ) ) {
		\Automattic\WooCommerce\Utilities\FeaturesUtil::declare_compatibility( 'custom_order_tables', __FILE__, false );
	}
} );

```


**Actual testing**
1. After completing pre-testing, let's go to WooCommerce > Settings> Advanced > Features. You should see the HPOS setting right on this screen.
2. Check that option to enable HPOS is disabled with the warning about incompatible plugins and unsynced order.
3. Start the sync, and test that it completes after some time.
4. Once sync is complete, go to the screen now, the warning about unsynced orders should go away.
5. Go the plugins screen and disable the test plugin that we uploaded.
6. Come back to the WooCommerce > Settings> Advanced > Features, option to enable HPOS should now be available.
7. Test that you can enable HPOS, go to the orders page and check that orders are listed as expected.

Please also refer to additional scenarios covered [in this comment](https://github.com/woocommerce/woocommerce/pull/38993#issuecomment-1630964276).

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [x] Enhancement

#### Message <!-- Add a changelog message here -->

Refresh the HPOS enable UX flow for better clarity.

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

vedanshujain committed:
Pass the correct param in woocommerce_settings_features.

vedanshujain committed:
Add initial layout for new HPOS setting.

vedanshujain committed:
Add changelog.

vedanshujain committed:
Some more fixups to refreshed UX.

vedanshujain committed:
Use inbuilt rendering block instead of writing a new template.

vedanshujain committed:
Add name attr so that it displays correctly in plugins incompat page.

vedanshujain committed:
Remove hook for feature setting screen and transaction level setting.

vedanshujain committed:
Mark HPOS as always enabled.

vedanshujain committed:
Remove auto flip feature as its risky to flip tables without merchant interaction.

vedanshujain committed:
Better function names.

vedanshujain committed:
Fix options for back compat and readibility.

vedanshujain committed:
Return appropriate setting if WC is installing.

vedanshujain committed:
Reduce the scope of transaction and enable by default for all.

vedanshujain committed:
Coding standard fixes

vedanshujain committed:
Commit after flipping HPOS toggle to prevent PHPUnit from calling rollback.

vedanshujain committed:
More resilient unit tests + return errors as expected from fetching functions.

vedanshujain committed:
More unit test fixes + throw error when transaction level is invalid.

vedanshujain committed:
Temporary remove COMMIT to see if it fixes tests.

vedanshujain committed:
Remove expectation of HPOS setting in advance test.

vedanshujain committed:
Flush cache since we are updating option.

vedanshujain committed:
Fixups when enabling/disabling HPOS.

vedanshujain committed:
Change method name to be more descriptive.

vedanshujain committed:
Fix notice.

vedanshujain committed:
Mior fixups with enable permissioning.

vedanshujain committed:
phpcs fix.

barryhughes commented:
💡 I modified the test plugin slightly to avoid an error (`$compatibility` was not in the closure's scope).

barryhughes commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -109,11 +113,16 @@ public function __construct() {
-				'disable_ui'      => false,
+			// Options for HPOS features are added in CustomOrdersTableController to keep the logic in same place.
+			'custom_order_tables'  => array( // This exists for back-compat, otherwise is always enabled and hidden from WC 8.0.
+				'name'               => '',
```
Shall we give it a name?

```suggestion
				'name'               => __( 'High-Performance order storage (COT)', 'woocommerce' ),
```

Could also add back the description unless it is no longer used, but the name is definitely needed in some parts of the UI:

<div align="center"><img width="800" alt="Screenshot 2023-07-10 at 11 41 15" src="https://github.com/woocommerce/woocommerce/assets/3594411/728e3952-d6b0-4e20-a64c-eb69380039b9"></div>

barryhughes commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -606,5 +428,101 @@ private function register_post_type_for_order_placeholders(): void {
+			'title'       => __( 'Data storage for orders', 'woocommerce' ),
+			'type'        => 'radio',
+			'options'     => array(
+				'no'  => __( 'WordPress Post tables', 'woocommerce' ),
```
This is a nit, but I don't think `Post` should be capitalized (and it doesn't tend to be in other places where it's used in a sentence or phrase within the WP UI):

```suggestion
				'no'  => __( 'WordPress post tables', 'woocommerce' ),
```

barryhughes commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -606,5 +428,101 @@ private function register_post_type_for_order_placeholders(): void {
+		} elseif ( $sync_status['current_pending_count'] > 0 ) {
+			$sync_message = sprintf(
+				// translators: %d: number of pending orders.
+				__( 'Sync %d pending orders. You can switch data storage for orders only when posts and orders table are in sync.', 'woocommerce' ),
```
Not essential, but we could do:

```suggestion
				_n( 
					'Sync %d pending order. You can switch data storage for orders only when posts and orders table are in sync.', 'woocommerce' ),
					'Sync %d pending orders. You can switch data storage for orders only when posts and orders table are in sync.', 'woocommerce' ),
					$sync_status['current_pending_count'],
					'woocommerce'
				),
```

Or even just:

```suggestion
				__( 'Sync %d pending order(s). You can switch data storage for orders only when posts and orders table are in sync.', 'woocommerce' ),
```

vedanshujain replied:
Great suggestion!

barryhughes commented on plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:
```diff
@@ -606,5 +428,101 @@ private function register_post_type_for_order_placeholders(): void {
+			'id'       => DataSynchronizer::ORDERS_DATA_SYNC_ENABLED_OPTION,
+			'title'    => '',
+			'type'     => 'checkbox',
+			'desc'     => __( 'Keep posts and orders table in sync. (Compatibility mode)', 'woocommerce' ),
```
Another nit, sorry 😬 ... Grammarly seems to prefer:

```suggestion
			'desc'     => __( 'Keep the posts and orders tables in sync (compatibility mode).', 'woocommerce' ),
```

barryhughes commented on plugins/woocommerce/src/Utilities/PluginUtil.php:
```diff
@@ -171,4 +171,61 @@ private function handle_plugin_de_activation(): void {
+			} else {
+
+				$feature_warning = sprintf(
+				/* translators: %1\$s, %2\$s = printable plugin names, %3\$d = plugins count */
```
Indentation could be tweaked:

```suggestion
					/* translators: %1\$s, %2\$s = printable plugin names, %3\$d = plugins count */
```

barryhughes commented on plugins/woocommerce/src/Utilities/PluginUtil.php:
```diff
@@ -171,4 +171,61 @@ private function handle_plugin_de_activation(): void {
+				admin_url( 'plugins.php' )
+			);
+			/* translators: %s = URL of the plugins page */
+			$extra_desc_tip = sprintf( __( "<br><a href='%s'>View and manage</a>", 'woocommerce' ), $incompatible_plugins_url );
```
Thinking it is better to keep HTML out of translatable strings, what do you think of something like:

```suggestion
			/* translators: %1$s opening link tag %2$s closing link tag. */
			$extra_desc_tip = '<br>' . sprintf( 
				__( '%1$sView and manage%2$s', 'woocommerce' ), 
				'<a href="' . esc_url( $incompatible_plugins_url ) . '">',
				'</a>'
			);
```

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-settings.php:
```diff
@@ -439,7 +440,11 @@ class="<?php echo esc_attr( $value['class'] ); ?>"
-									<?php echo $description; // WPCS: XSS ok. ?>
+									<?php
+									if ( ! $show_desc_at_end ) {
+										echo wp_kses_post( $description );
```
(1/2) If the description is shown at the start, we echo `$description`.

barryhughes commented on plugins/woocommerce/includes/admin/class-wc-admin-settings.php:
```diff
@@ -458,6 +463,9 @@ class="<?php echo esc_attr( $value['class'] ); ?>"
 										<?php
 									}
+									if ( $show_desc_at_end ) {
+										echo wp_kses_post( "<p class='description description-thin'>{$value['desc']}</p>" );
```
(2/2) If the description is shown at the end, we echo `$value['desc']`. Even if these are effectively aliases, seems it might be better to use the same one in both places?

barryhughes replied:
Unless this is deliberate, of course.

vedanshujain replied:
I used `value['desc']` because `$description` won't be set when `desc_tip` is true, but on thinking further, it looks like it's fine not to set it in those cases. Added the recommendation.

barryhughes reviewed (commented):
Sharing some straightforward and minor notes on the UI part of this change, still looking at the other half (changes to migration code, etc).

barryhughes reviewed (commented):
Nice!

Beyond the small things noted earlier, this tests well and I was able to re-sync and view orders post-migration, as expected.

vedanshujain committed:
Add name so that it's displayed in the incompatible plugins UX.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

vedanshujain committed:
Capitalisation fix for posts.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

vedanshujain committed:
Use proper pluralization of pending sync message.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

vedanshujain committed:
Add the `the`.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

vedanshujain committed:
Fix indentation.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

vedanshujain committed:
Move out HTML out of translatable string for easier translation.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

vedanshujain committed:
Fix syntax from GH UX commit + use consistent description.

vedanshujain commented:
To test

**On a new shop, with at least one incompatible plugin but no orders**
- [ ] When installing WC for the first time, posts are enabled. Additionally, there is no warning even when installing incompatible plugins (since post tables are enabled).
- [ ] On the setting page, the option to switch to HPOS is disabled. However, the option to keep posts and HPOS in sync is enabled.
- [ ] In the settings page, the incompatible plugin is listed. Additionally, clicking on "View and manage" takes to the plugin list page with incompatible plugins visible.
- [ ] We show correct tools visibility in WooCommerce > Status > Tools. The tool for creating order tables should be visible, and deleting order tables should be hidden.
- [ ] Deactivating the incompatible plugin, enables the option to switch to HPOS (assuming no orders).
- [ ] Enabling the HPOS also creates the order tables.
- [ ] Tools visibility should be inverted, now the tool for dropping the order table should be visible. However, it is still disabled, since HPOS is authoritative.
- [ ] Option to switch back to the post should be available. When switched back, the option to drop the order table should become visible again.

**On a shop with existing orders and post authoritative with all compatible plugins is updated to a version that contains this PR (custom order feature was not enabled)**
- [ ] Posts are enabled by default, but the option to switch to HPOS is disabled. The radio box for posts is also disabled indicating that till there are orders to be synced, the authoritative source cannot be changed.
- [ ] On enabling the option to keep posts and HPOS in sync, the message should be displayed in the settings section that sync is currently in progress.
- [ ] At this point, going to WooCommerce > Status > Scheduled Actions should display the action pending for sync in either the pending or running tab.
- [ ] After a while sync should complete, or at least the count should start decreasing.
- [ ] Option to turn off sync should always be available.
- [ ] When turned off, the action for migrating orders will not be scheduled again, even if there are pending orders. Already scheduled action will still be processed.
- [ ] Tools visibility is as expected similar to the new shop flow.

**Other misc testings**
- [ ] On shop with HPOS already enabled, switching to this PR will still have all corresponding settings enabled.
- [ ] On shop with MyISAM table instead of InnoDB, transaction commands still works as expected without locking the orders table.

vedanshujain committed:
Coding standard fixups.

barryhughes commented:
**On a new shop, with at least one incompatible plugin but no orders**

- [x] When installing WC for the first time, posts are enabled. Additionally, there is no warning even when installing incompatible plugins (since post tables are enabled).
- [ ] On the setting page, the option to switch to HPOS is disabled. However, the option to keep posts and HPOS in sync is enabled.
    - **No** → On a brand new site, after activating WooCommerce, the **Keep the posts and orders tables in sync (compatibility mode)** option is not checked.
- [x] In the settings page, the incompatible plugin is listed. Additionally, clicking on "View and manage" takes to the plugin list page with incompatible plugins visible.
    - Yes, once activated.
- [ ] We show correct tools visibility in WooCommerce > Status > Tools. The tool for creating order tables should be visible, and deleting order tables should be hidden.
    - **No** → The tool for deleting order tables is visible, and the tables are already in existence.
    - I wiped my install and started again to be sure; `<prefix>_wc_orders` is created as soon as I run through NUX (but probably earlier). Ditto the order meta table.
- [ ] Deactivating the incompatible plugin, enables the option to switch to HPOS (assuming no orders).
- [ ] Enabling the HPOS also creates the order tables.
- [ ] Tools visibility should be inverted, now the tool for dropping the order table should be visible. However, it is still disabled, since HPOS is authoritative.
- [ ] Option to switch back to the post should be available. When switched back, the option to drop the order table should become visible again.

barryhughes commented:
Sorry for the delay, @vedanshujain. Left notes on the initial set of testing steps: do you want to check those out before I proceed to the second and third sets?

vedanshujain committed:
Use custom option for custom order tables feature.

vedanshujain committed:
Set option when HPOS tables are created.

Earlier, we were marking this based on whether HPOS is enabled or not, but now that HPOS is enabled (but not authoritative) for everyone, we need a seperate option to track this.

vedanshujain committed:
Remove references to removed COT seperate feature flag.

vedanshujain committed:
Handle edge case of order table being unexpectedly not present.

vedanshujain commented:
@barryhughes this is ready for review again, but note that I have not yet implemented the language softening we chatted about yesterday. Let's address that in a separate PR, since this one is getting too big, and I want to spread the test build around soon-ish. wdyt?

barryhughes commented:
> Let's address that in a separate PR, since this one is getting too big, and I want to spread the test build around soon-ish. wdyt?

:+1:

barryhughes commented on plugins/woocommerce/src/Database/Migrations/CustomOrderTable/PostsToOrdersMigrationController.php:
```diff
@@ -26,7 +26,7 @@ class PostsToOrdersMigrationController {
 	 * Array of objects used to perform the migration.
 	 *
-	 * @var array
+	 * @var array[TableMigrator]
```
Feel free to ignore me if I'm wrong—it just jumped out because I don't think I've often seen it expressed this way—should this be:

```suggestion
	 * @var TableMigrator[]
```

https://docs.phpdoc.org/2.9/references/phpdoc/types.html#arrays

barryhughes commented on plugins/woocommerce/src/Database/Migrations/MetaToMetaTableMigrator.php:
```diff
@@ -61,33 +61,77 @@ public function __construct() {
 		$to_migrate = $this->fetch_data_for_migration_for_ids( $entity_ids );
 		if ( empty( $to_migrate ) ) {
-			return;
+			return array(
+				array(),
+				array(),
+			);
```
Also optional, because I think wherever the returned array is consumed we have appropriate checks, but, should this ideally be:

```suggestion
			return array(
				'data'   => array(),
				'errors' => array(),
			);
```

vedanshujain replied:
Thanks for catching, seems like I missed this.

vedanshujain replied:
Addressed in https://github.com/woocommerce/woocommerce/pull/38993/commits/94fce9c169809e484a9d75803600ae3ad88b4ee5

barryhughes commented on plugins/woocommerce/src/Internal/Features/FeaturesController.php:
```diff
@@ -722,16 +703,23 @@ private function get_setting_for_feature( string $feature_id, array $feature, bo
+		/**
+		 * Allows to modify feature setting that will be used to render in the feature page.
+		 *
+		 * @since 8.0.0
```
Should we document the filter params?

```suggestion
		 * @since 8.0.0
		 *
		 * @param array $feature_setting {
		 *     Describes the feature.
		 *
		 *     @type string $title    Feature title.
		 *     @type string $desc     Feature description.
		 *     @type string $type     Example: 'checkbox'.
		 *     @type string $id       Feature ID/slug.
		 *     @type bool   $disabled If disabled.
		 *     @type string $desc_tip Description tooltip text.
		 *     @type string $default  Default value for this feature setting.
		 * }
		 * @param string $feature_id  The feature ID.
		 
```

vedanshujain replied:
Addressed in https://github.com/woocommerce/woocommerce/pull/38993/commits/94fce9c169809e484a9d75803600ae3ad88b4ee5

barryhughes commented on plugins/woocommerce/src/Utilities/PluginUtil.php:
```diff
@@ -171,4 +171,65 @@ private function handle_plugin_de_activation(): void {
+				$feature_warning = sprintf( __( '⚠ 1 Incompatible plugin detected (%s).', 'woocommerce' ), $this->get_plugin_name( $incompatibles[0] ) );
+			} elseif ( 2 === $incompatible_count ) {
+				$feature_warning = sprintf(
+				/* translators: %1\$s, %2\$s = printable plugin names */
```
```suggestion
					/* translators: %1\$s, %2\$s = printable plugin names */
```

vedanshujain replied:
Addressed in https://github.com/woocommerce/woocommerce/pull/38993/commits/94fce9c169809e484a9d75803600ae3ad88b4ee5

barryhughes reviewed (approved):
Great! Some optional but non-required minor changes. This feels really nice.

vedanshujain committed:
Use correct representation of array types.

Co-authored-by: Barry Hughes <3594411+barryhughes@users.noreply.github.com>

vedanshujain committed:
Apply PR feedback suggestions.

vedanshujain committed:
Convert tabs to spaces.
